README.md
# NAME

UV::Util - Some utility functions from libUV.

# WARNING

UV::Util is being released as a learning process that will hopefully be folded
into a more complete set of modules allowing use of libUV with Perl later. In
other words, this will be more useful later. If it happens to be useful now,
that's awesome. If not, hopefully it will be soon. Until then, please let me know
of any errors during installation or of anything I can do to make this better.

# SYNOPSIS

    #!/usr/bin/env perl
    use strict;
    use warnings;
    use feature ':5.14';

    use Data::Dumper::Concise qw(Dumper);
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try {
      $res = UV::Util::cpu_info();
    }
    catch {
      die "Aw, man. $@";
    }

# DESCRIPTION

This module provides access to a few of the functions in the miscellaneous
[libUV](http://docs.libuv.org/en/v1.x/misc.html) utilities. While it's extremely
unlikely, all functions here can throw an exception on error unless specifically
stated otherwise in the function's description.

# CONSTANTS

[UV::Util](https://metacpan.org/pod/UV::Util) makes the following constants available that represent different
[handle types](http://docs.libuv.org/en/latest/handle.html#c.uv_handle_type).

## UV\_UNKNOWN\_HANDLE

## UV\_ASYNC

## UV\_CHECK

## UV\_FS\_EVENT

## UV\_FS\_POLL

## UV\_HANDLE

## UV\_IDLE

## UV\_NAMED\_PIPE

## UV\_POLL

## UV\_PREPARE

## UV\_PROCESS

## UV\_STREAM

## UV\_TCP

## UV\_TIMER

## UV\_TTY

## UV\_UDP

## UV\_SIGNAL

## UV\_FILE

## UV\_HANDLE\_TYPE\_MAX

# FUNCTIONS

All functions provided here provide a Perl interface to their
[libUV](http://docs.libuv.org/en/latest/misc.html) equivalents.

## cpu\_info

    use Data::Dumper;
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::cpu_info(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res;
    # [
    #   {
    #     cpu_times => {
    #       idle => "1157161200",
    #       irq => 19912800,
    #       nice => 242400,
    #       sys => 8498700,
    #       user => 39428000
    #     },
    #     model => "Intel(R) Core(TM) i7-7700HQ CPU \@ 2.80GHz",
    #     speed => 1048
    #   },
    #   {
    #     cpu_times => {
    #       idle => 16922900,
    #       irq => 846300,
    #       nice => 371800,
    #       sys => 6491400,
    #       user => 34690000
    #     },
    #     model => "Intel(R) Core(TM) i7-7700HQ CPU \@ 2.80GHz",
    #     speed => 899
    #   }
    # ]

This [function](http://docs.libuv.org/en/v1.x/misc.html#c.uv_cpu_info)
returns an array reference full of hashrefs. Each hashref represents an
available CPU on your system. `cpu_times`, `model`, and `speed` will be
supplied for each.

## get\_free\_memory

    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::get_free_memory(); }
    catch { die "Aw, man. $@"; }
    say $res; # 23052402688

This function returns an unsigned integer representing the number of bytes of
free memory available.

## get\_total\_memory

    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::get_total_memory(); }
    catch { die "Aw, man. $@"; }
    say $res; # 33452101632

This function returns an unsigned integer representing the number of bytes of
total memory in the system.

## getrusage

    use Data::Dumper;
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::getrusage(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res;
    # {
    #   ru_idrss => 0,
    #   ru_inblock => 0,
    #   ru_isrss => 0,
    #   ru_ixrss => 0,
    #   ru_majflt => 0,
    #   ru_maxrss => 10132,
    #   ru_minflt => 1624,
    #   ru_msgrcv => 0,
    #   ru_msgsnd => 0,
    #   ru_nivcsw => 1,
    #   ru_nsignals => 0,
    #   ru_nswap => 0,
    #   ru_nvcsw => 1,
    #   ru_oublock => 0,
    #   ru_stime => "0.005963",
    #   ru_utime => "0.02801"
    # }

This [function](http://docs.libuv.org/en/v1.x/misc.html#c.uv_getrusage)
returns a hash reference of resource metrics for the current process.

## guess\_handle\_type

    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::guess_handle_type(\*STDIN); }
    catch { die "Aw, man. $@"; }
    say "yay!" if ($res == UV::Util::UV_TTY);

This [function](http://docs.libuv.org/en/latest/misc.html#c.uv_guess_handle)
takes in a reference to a handle (e.g. `\*STDIN`) and returns an integer that
represents one of the [CONSTANTS](https://metacpan.org/pod/UV::Util#CONSTANTS) above.

## hrtime

    use UV::Util;

    # does not throw exceptions
    my $time = UV::Util::hrtime();

This [function](http://docs.libuv.org/en/latest/misc.html#c.uv_hrtime)
returns the current high-resolution real time. This is expressed in nanoseconds.
It is relative to an arbitrary time in the past. It is not related to the time
of day and therefore not subject to clock drift. The primary use is for
measuring performance between intervals.

Not every platform can support nanosecond resolution; however, this value will
always be in nanoseconds.

## interface\_addresses

    use Data::Dumper qw(Dumper);
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::interface_addresses(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res;
    # [
    #   {
    #     address => "127.0.0.1",
    #     is_internal => 1,
    #     mac => "00:00:00:00:00:00",
    #     name => "lo",
    #     netmask => "255.0.0.0"
    #   },
    # ]

This [function](http://docs.libuv.org/en/latest/misc.html#c.uv_interface_addresses)
returns an array reference containing hash references representing each of your
available interfaces.

## loadavg

    use Data::Dumper qw(Dumper);
    use UV::Util;

    # does not throw exceptions
    my $res = UV::Util::loadavg();
    say Dumper $res;
    # [
    #   "0.43212890625",
    #   "0.39599609375",
    #   "0.27880859375"
    # ]

This [function](http://docs.libuv.org/en/latest/misc.html#c.uv_loadavg)
returns an array reference containing the
[load average](http://en.wikipedia.org/wiki/Load_\(computing\)).

On Windows, this will **always** return `[0,0,0]` as it's not implemented.

## resident\_set\_memory

    use Data::Dumper qw(Dumper);
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::resident_set_memory(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res; # 10473472

This [function](http://docs.libuv.org/en/latest/misc.html#c.uv_resident_set_memory)
returns an unsigned integer representing the resident set size (RSS) for the
current process.

## uptime

    use Data::Dumper qw(Dumper);
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::uptime(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res; # 603468

This [function](http://docs.libuv.org/en/latest/misc.html#c.uv_uptime)
returns a float representing the current system uptime.

## version

    use UV::Util;

    say UV::Util::version();
    # 1.9.1

This [function](http://docs.libuv.org/en/v1.x/version.html)
returns the libUV version number as a string. Does not throw errors.

# COPYRIGHT AND LICENSE

Copyright 2017, Chase Whitener.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.


Changes
Revision history for Perl extension {{$dist->name}}

{{$NEXT}}

1.000     2018-04-05
    - Update to use the cleaned up version of Alien::libuv
    - Remove the build complexity for windows as that's in Alien::libuv now
    - Bump the major version to 1.000 to coincide with the Alien::libuv change.

0.010     2018-04-03
    - Update the tests to pass on cygwin.
    - Fix unnecessarily added libraries during build on cygwin
    - Fixup AppVeyor tests some more to expand testing.

0.009     2018-03-08
    - Add AppVeyor tests
    - Move constants out in the same way as with UV

0.008     2017-07-10
    - Bump version of Alien::libuv to 0.006 to add more diagnostics
    - Add more diagnostics to the tests

0.007     2017-06-29
    - Bump the requirement of Alien::libuv to v0.004 to ensure libuv v1+

0.006     2017-06-27
    - Add Exporter to prereqs.

0.005     2017-06-27
    - Make all functions and constants exportable

0.004     2017-06-27
    - Bump requirement for Alien::libuv to 0.003, ensuring clean installs on
      more platforms
    - Add PROTOTYPES: DISABLE
    - Add XSLoader to the prereqs list

0.003     2017-06-16
    - Using DZil now
    - Make use of Alien::libuv

0.002  2017-06-10
    - Removed some stupid assumptions about compiler flags (thanks, mauke)
    - Added a version function to return the version of libUV installed

0.001  2017-06-08
    - Implemented and documented the UV::Util functions


Makefile.PL
# This Makefile.PL for UV-Util was generated by
# Dist::Zilla::Plugin::MakeMaker::Awesome 0.41.
# Don't edit it but the dist.ini and plugins used to construct it.

use strict;
use warnings;

use 5.008001;
use ExtUtils::MakeMaker;

use Config;
use Alien::libuv ();

# make sure we actually get stuff back from Alien::libuv
sub TRIM {
    my $s = shift;
    return '' unless $s;
    $s =~ s/\A\s*//;
    $s =~ s/\s*\z//;
    return $s;
}

my @flags = ('-I.', $Config{ccflags});
my $libs = Alien::libuv->libs();
{
    my $cflags = TRIM(Alien::libuv->cflags);
    my $cflags_s = TRIM(Alien::libuv->cflags_static);
    if ($cflags eq $cflags_s) {
        unshift @flags, $cflags if $cflags;
    }
    else {
        unshift @flags, $cflags if $cflags;
        unshift @flags, $cflags_s if $cflags_s;
    }
}

# this will get merged into the meta
my %xsbuild = (
    XSMULTI => 1,
    LIBS => [$libs],
    CCFLAGS => join(' ', @flags),
);

my %WriteMakefileArgs = (
  "ABSTRACT" => "Some utility functions from libUV.",
  "AUTHOR" => "Chase Whitener <capoeirab\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Alien::libuv" => "1.000",
    "Config" => 0,
    "ExtUtils::MakeMaker" => "7.12"
  },
  "CONFIGURE_REQUIRES" => {
    "Alien::libuv" => 0,
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "UV-Util",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "UV::Util",
  "PREREQ_PM" => {
    "Alien::libuv" => "1.000",
    "Exporter" => "5.57",
    "XSLoader" => "0.14",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Alien::libuv" => "1.000",
    "ExtUtils::MakeMaker" => "7.12",
    "File::Spec" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "1.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);

%WriteMakefileArgs = (
    %WriteMakefileArgs,
    %xsbuild,
);

my %FallbackPrereqs = (
  "Alien::libuv" => "1.000",
  "Config" => 0,
  "Exporter" => "5.57",
  "ExtUtils::MakeMaker" => "7.12",
  "File::Spec" => 0,
  "Test::More" => "0.88",
  "XSLoader" => "0.14",
  "strict" => 0,
  "warnings" => 0
);

unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/UV/Util.pm
package UV::Util;

our $VERSION = '1.001';
our $XS_VERSION = $VERSION;
$VERSION = eval $VERSION;

use strict;
use warnings;
use Exporter qw(import);
require XSLoader;

our @EXPORT_OK  = (@EXPORT_OK, qw(
    cpu_info get_free_memory get_total_memory getrusage guess_handle_type
    hrtime interface_addresses loadavg resident_set_memory version
));

XSLoader::load('UV::Util', $XS_VERSION);

1;
__END__

=encoding utf8

=head1 NAME

UV::Util - Some utility functions from libUV.

=head1 WARNING

UV::Util is being released as a learning process that will hopefully be folded
into a more complete set of modules allowing use of libUV with Perl later. In
other words, this will be more useful later. If it happens to be useful now,
that's awesome. If not, hopefully it will be soon. Until then, please let me know
of any errors during installation or of anything I can do to make this better.

=head1 SYNOPSIS

  #!/usr/bin/env perl
  use strict;
  use warnings;
  use feature ':5.14';

  use Data::Dumper::Concise qw(Dumper);
  use UV::Util;
  use Syntax::Keyword::Try;

  my $res;
  try {
    $res = UV::Util::cpu_info();
  }
  catch {
    die "Aw, man. $@";
  }

=head1 DESCRIPTION

This module provides access to a few of the functions in the miscellaneous
L<libUV|http://docs.libuv.org/en/v1.x/misc.html> utilities. While it's extremely
unlikely, all functions here can throw an exception on error unless specifically
stated otherwise in the function's description.

=head1 CONSTANTS

L<UV::Util> makes the following constants available that represent different
L<handle types|http://docs.libuv.org/en/latest/handle.html#c.uv_handle_type>.

=head2 UV_UNKNOWN_HANDLE

=head2 UV_ASYNC

=head2 UV_CHECK

=head2 UV_FS_EVENT

=head2 UV_FS_POLL

=head2 UV_HANDLE

=head2 UV_IDLE

=head2 UV_NAMED_PIPE

=head2 UV_POLL

=head2 UV_PREPARE

=head2 UV_PROCESS

=head2 UV_STREAM

=head2 UV_TCP

=head2 UV_TIMER

=head2 UV_TTY

=head2 UV_UDP

=head2 UV_SIGNAL

=head2 UV_FILE

=head2 UV_HANDLE_TYPE_MAX

=head1 FUNCTIONS

All functions provided here provide a Perl interface to their
L<libUV|http://docs.libuv.org/en/latest/misc.html> equivalents.

=head2 cpu_info

    use Data::Dumper;
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::cpu_info(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res;
    # [
    #   {
    #     cpu_times => {
    #       idle => "1157161200",
    #       irq => 19912800,
    #       nice => 242400,
    #       sys => 8498700,
    #       user => 39428000
    #     },
    #     model => "Intel(R) Core(TM) i7-7700HQ CPU \@ 2.80GHz",
    #     speed => 1048
    #   },
    #   {
    #     cpu_times => {
    #       idle => 16922900,
    #       irq => 846300,
    #       nice => 371800,
    #       sys => 6491400,
    #       user => 34690000
    #     },
    #     model => "Intel(R) Core(TM) i7-7700HQ CPU \@ 2.80GHz",
    #     speed => 899
    #   }
    # ]

This L<function|http://docs.libuv.org/en/v1.x/misc.html#c.uv_cpu_info>
returns an array reference full of hashrefs. Each hashref represents an
available CPU on your system. C<cpu_times>, C<model>, and C<speed> will be
supplied for each.

=head2 get_free_memory

    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::get_free_memory(); }
    catch { die "Aw, man. $@"; }
    say $res; # 23052402688

This function returns an unsigned integer representing the number of bytes of
free memory available.

=head2 get_total_memory

    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::get_total_memory(); }
    catch { die "Aw, man. $@"; }
    say $res; # 33452101632

This function returns an unsigned integer representing the number of bytes of
total memory in the system.

=head2 getrusage

    use Data::Dumper;
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::getrusage(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res;
    # {
    #   ru_idrss => 0,
    #   ru_inblock => 0,
    #   ru_isrss => 0,
    #   ru_ixrss => 0,
    #   ru_majflt => 0,
    #   ru_maxrss => 10132,
    #   ru_minflt => 1624,
    #   ru_msgrcv => 0,
    #   ru_msgsnd => 0,
    #   ru_nivcsw => 1,
    #   ru_nsignals => 0,
    #   ru_nswap => 0,
    #   ru_nvcsw => 1,
    #   ru_oublock => 0,
    #   ru_stime => "0.005963",
    #   ru_utime => "0.02801"
    # }

This L<function|http://docs.libuv.org/en/v1.x/misc.html#c.uv_getrusage>
returns a hash reference of resource metrics for the current process.

=head2 guess_handle_type

    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::guess_handle_type(\*STDIN); }
    catch { die "Aw, man. $@"; }
    say "yay!" if ($res == UV::Util::UV_TTY);

This L<function|http://docs.libuv.org/en/latest/misc.html#c.uv_guess_handle>
takes in a reference to a handle (e.g. C<\*STDIN>) and returns an integer that
represents one of the L<CONSTANTS|UV::Util/CONSTANTS> above.

=head2 hrtime

    use UV::Util;

    # does not throw exceptions
    my $time = UV::Util::hrtime();

This L<function|http://docs.libuv.org/en/latest/misc.html#c.uv_hrtime>
returns the current high-resolution real time. This is expressed in nanoseconds.
It is relative to an arbitrary time in the past. It is not related to the time
of day and therefore not subject to clock drift. The primary use is for
measuring performance between intervals.

Not every platform can support nanosecond resolution; however, this value will
always be in nanoseconds.

=head2 interface_addresses

    use Data::Dumper qw(Dumper);
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::interface_addresses(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res;
    # [
    #   {
    #     address => "127.0.0.1",
    #     is_internal => 1,
    #     mac => "00:00:00:00:00:00",
    #     name => "lo",
    #     netmask => "255.0.0.0"
    #   },
    # ]


This L<function|http://docs.libuv.org/en/latest/misc.html#c.uv_interface_addresses>
returns an array reference containing hash references representing each of your
available interfaces.

=head2 loadavg

    use Data::Dumper qw(Dumper);
    use UV::Util;

    # does not throw exceptions
    my $res = UV::Util::loadavg();
    say Dumper $res;
    # [
    #   "0.43212890625",
    #   "0.39599609375",
    #   "0.27880859375"
    # ]

This L<function|http://docs.libuv.org/en/latest/misc.html#c.uv_loadavg>
returns an array reference containing the
L<load average|http://en.wikipedia.org/wiki/Load_(computing)>.

On Windows, this will B<always> return C<[0,0,0]> as it's not implemented.

=head2 resident_set_memory

    use Data::Dumper qw(Dumper);
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::resident_set_memory(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res; # 10473472

This L<function|http://docs.libuv.org/en/latest/misc.html#c.uv_resident_set_memory>
returns an unsigned integer representing the resident set size (RSS) for the
current process.

=head2 uptime

    use Data::Dumper qw(Dumper);
    use UV::Util;
    use Syntax::Keyword::Try;

    my $res;
    try { $res = UV::Util::uptime(); }
    catch { die "Aw, man. $@"; }
    say Dumper $res; # 603468

This L<function|http://docs.libuv.org/en/latest/misc.html#c.uv_uptime>
returns a float representing the current system uptime.

=head2 version

    use UV::Util;

    say UV::Util::version();
    # 1.9.1

This L<function|http://docs.libuv.org/en/v1.x/version.html>
returns the libUV version number as a string. Does not throw errors.

=head1 COPYRIGHT AND LICENSE

Copyright 2017, Chase Whitener.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
